File: 백준/Gold/3190. 뱀/뱀.py
from sys import stdin as s
from collections import deque

#s = open('input.txt','rt')
snake = deque([])
N = int(s.readline())
K = int(s.readline())

arr = [[0] * N for _ in range(N)]

for i in range(K):
    X,Y = map(int,s.readline().split())
    arr[X-1][Y-1]=1
    
time_list =deque([])
L = int(s.readline())

for i in range(L):
    time_set, vector = list(map(str,s.readline().split()))
    time_list.append((time_set,vector))



# 동, 남, 서, 북 (시계방향)
dx = [0,1,0,-1]
dy = [1,0,-1,0]

# 최초 값 설정
x,y,d = 0,0,0
time = 0

while True:
    time +=1
    snake.append((x,y))
    
    x += dx[d]
    y += dy[d]
        
    if x<0 or x>=N or y<0 or y>=N or arr[x][y]==2:
        break
    
    if not arr[x][y]:
        i,j = snake.popleft()
        arr[i][j] =0
        
    arr[x][y] =2
    
    if time_list:
        if time == int(time_list[0][0]):
            if time_list[0][1] =='D':  #D일경우 오른쪽으로 90도 회전
                d = (d+1) % 4
                time_list.popleft() 
            else:                      #D가 아닐경우(L) 왼쪽으로 90도 회전
                d = (d-1) % 4
                time_list.popleft()
        
     
print(time)

File: 백준/Gold/3190. 뱀/뱀.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Deque;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.Map;
import java.util.StringTokenizer;

public class Main {
	static int N;
	static int[][] board;
	static Map<Integer, String> directionMap = new HashMap<>();
	static int[] dx = {0, 1, 0, -1};
	static int[] dy = {1, 0, -1, 0};
	public static void main(String[] args) throws IOException {

		// BufferedReader br = new BufferedReader(new FileReader("src/input.txt"));
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		N = Integer.parseInt(br.readLine());
		int K = Integer.parseInt(br.readLine());
		board = new int[N][N];

		for (int i = 0; i < K; i++) {
			StringTokenizer st = new StringTokenizer(br.readLine());
			int x = Integer.parseInt(st.nextToken()) - 1;
			int y = Integer.parseInt(st.nextToken()) - 1;
			board[x][y] = 1;
		}

		int L = Integer.parseInt(br.readLine());
		for (int i = 0; i < L; i++) {
			StringTokenizer st = new StringTokenizer(br.readLine());
			int time = Integer.parseInt(st.nextToken());
			String direction = st.nextToken();
			directionMap.put(time, direction);
		}
		System.out.println(simulate());
	}

	static int simulate() {
		Deque<int[]> snake = new LinkedList<>();
		snake.add(new int[] {0, 0});
		int time=0, direction=0;

		while (!snake.isEmpty()) {
			time++;
			int[] head = snake.peekLast();
			int nx = head[0] + dx[direction];
			int ny = head[1] + dy[direction];

			if (nx < 0 || nx >=N || ny < 0 || ny >=N || contains(snake, nx, ny)) {
				break;
			}

			if (board[nx][ny] == 1) {
				board[nx][ny] = 0;
			} else {
				snake.pollFirst();
			}

			snake.addLast(new int[] {nx, ny});

			if (directionMap.containsKey(time)) {
				direction = changeDirection(direction, directionMap.get(time));
			}
		}
		return time;
	}

	static int changeDirection(int X, String C) {
		if (C.equals("L")) {
			return(X+3)%4;
		} else {
			return(X+1)%4;
		}
	}

	static boolean contains(Deque<int[]> snake, int nx, int ny) {
		for (int[] part : snake) {
			if (part[0] == nx && part[1] == ny) {
				return true;
			}
		}
		return false;
	}
}
